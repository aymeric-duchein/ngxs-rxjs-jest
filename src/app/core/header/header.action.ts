export class SignOutButtonClick {
  static readonly type = '[HEADER] Sign out button click';
}

export class AddIssueBtnClick {
  static readonly type = '[HEADER] Add issue button click';
}

export class AddHttpIssueBtnClick {
  static readonly type = '[HEADER] Add http issue button click';
}
