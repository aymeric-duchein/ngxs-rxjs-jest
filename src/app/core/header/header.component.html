<mat-toolbar color="primary">
  <span>Issue board</span>
  <ng-container *ngIf="isAuthenticated$ | async">
    <div class="header-buttons">
      <div>
        <a mat-button [routerLink]="routesNames.BOARD">
          <mat-icon>dashboard</mat-icon>
          Board</a>
        <a mat-button [routerLink]="routesNames.LISTING">
          <mat-icon>list</mat-icon>
          Listing</a>
      </div>
      <div>
        <button mat-flat-button color="primary" [matMenuTriggerFor]="menu">
          {{ userName$ | async }}
          <mat-icon>account_circle</mat-icon></button>
        <mat-menu #menu="matMenu">
          <button mat-menu-item [routerLink]="routesNames.SETTINGS">
            <mat-icon>settings_applications</mat-icon>
            <span>Settings</span>
          </button>
          <button mat-menu-item (click)="signOut()">
            <mat-icon>power_settings_new</mat-icon>
            <span>Sign out</span>
          </button>
        </mat-menu>
      </div>
    </div>
  </ng-container>

</mat-toolbar>
