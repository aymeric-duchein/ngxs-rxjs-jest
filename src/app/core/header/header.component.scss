mat-toolbar {
  justify-content: space-between;
}

.header-buttons {
  width: 100%;
  display: flex;
  justify-content: space-between;

  div {
    padding-left: 1rem;
  }
}

:host {
  width: 100vw;
}
