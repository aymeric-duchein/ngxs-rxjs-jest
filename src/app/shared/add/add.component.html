<mat-icon class="add-button" [matMenuTriggerFor]="menu">add_circle</mat-icon>
<mat-menu #menu="matMenu">
  <button mat-menu-item (click)="addIssue()">
    <mat-icon>note_add</mat-icon>
    <span>Add issue</span>
  </button>
  <button mat-menu-item (click)="addHttpIssue()">
    <mat-icon>cached</mat-icon>
    <span>Add random issue</span>
  </button>
</mat-menu>
