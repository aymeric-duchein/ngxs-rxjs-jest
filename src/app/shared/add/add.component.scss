@import "src/styles";

.add-button {
  height: 40px;
  width: 40px;
  font-size: 40px;
  color: #5A6773;
  cursor: pointer;
}



