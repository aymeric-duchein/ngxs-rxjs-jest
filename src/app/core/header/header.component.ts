import {Component} from '@angular/core';
import {Select, Store} from '@ngxs/store';
import {UserState} from '../user/user.state';
import {Observable} from 'rxjs';
import {RoutesNames} from '../../app-routing.const';
import {SignOutButtonClick} from './header.action';

@Component({
  selector: 'app-header',
  templateUrl: './header.component.html',
  styleUrls: ['./header.component.scss']
})
export class HeaderComponent {

  routesNames = RoutesNames;

  @Select(UserState.isAuthenticated) isAuthenticated$: Observable<boolean>;
  @Select(UserState.getUserName) userName$: Observable<string>;

  constructor(private readonly store: Store) {
  }

  signOut() {
    this.store.dispatch(new SignOutButtonClick());
  }
}
