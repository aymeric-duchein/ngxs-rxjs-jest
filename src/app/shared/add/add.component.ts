import {Component} from '@angular/core';
import {Store} from '@ngxs/store';
import {AddHttpIssueBtnClick, AddIssueBtnClick} from '../../core/header/header.action';

@Component({
  selector: 'app-add',
  templateUrl: './add.component.html',
  styleUrls: ['./add.component.scss']
})
export class AddComponent {
  constructor(private readonly store: Store) {
  }

  addIssue() {
    this.store.dispatch(new AddIssueBtnClick());
  }

  addHttpIssue() {
    this.store.dispatch(new AddHttpIssueBtnClick());
  }
}
