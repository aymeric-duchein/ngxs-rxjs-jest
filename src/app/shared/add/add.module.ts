import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {MatButtonModule, MatIconModule, MatMenuModule} from '@angular/material';
import {AddComponent} from './add.component';

@NgModule({
  declarations: [AddComponent],
  exports: [AddComponent],
  imports: [
    CommonModule,
    MatButtonModule,
    MatMenuModule,
    MatIconModule,
  ],
})
export class AddModule {
}
