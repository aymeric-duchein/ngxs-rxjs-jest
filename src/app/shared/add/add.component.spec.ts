import {AddComponent} from '@src/app/shared/add/add.component';
import {ComponentFixture, TestBed} from '@angular/core/testing';
import {NgxsModule, Store} from '@ngxs/store';
import {MatButtonModule, MatIconModule, MatMenuModule, MatMenuTrigger} from '@angular/material';
import {MockHelper, MockModule} from 'ng-mocks';
import {By} from '@angular/platform-browser';
import {AddHttpIssueBtnClick, AddIssueBtnClick} from '@src/app/core/header/header.action';


describe('AddComponent', () => {
  let component: AddComponent;
  let fixture: ComponentFixture<AddComponent>;
  let store: Store;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [AddComponent],
      imports: [
        NgxsModule.forRoot(),
        MockModule(MatButtonModule),
        MockModule(MatMenuModule),
        MockModule(MatIconModule),
      ]
    });
    store = TestBed.get(Store);
    jest.spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(AddComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  test('should trigger menu opening', () => {
    const icon = fixture.debugElement.query(By.css('mat-icon'));
    const menuTrigger = MockHelper.getDirective(icon, MatMenuTrigger);
    expect(menuTrigger).not.toBeNull();
  });

  test('should display two buttons', () => {
    const buttons = fixture.debugElement.queryAll(By.css('button'));
    expect(buttons.length).toEqual(2);
  });

  test('should add a new issue', () => {
    const addIssue = fixture.debugElement.query(By.css('button:nth-of-type(1)'));
    expect(addIssue).not.toBeNull();
    addIssue.triggerEventHandler('click', {});
    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(expect.any(AddIssueBtnClick));
    expect(store.dispatch).toHaveBeenCalledWith(new AddIssueBtnClick());
  });

  test('should add a new http issue', () => {
    const addHttpIssue = fixture.debugElement.query(By.css('button:nth-of-type(2)'));
    expect(addHttpIssue).not.toBeNull();
    addHttpIssue.triggerEventHandler('click', {});
    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(expect.any(AddHttpIssueBtnClick));
    expect(store.dispatch).toHaveBeenCalledWith(new AddHttpIssueBtnClick());
  });
});



