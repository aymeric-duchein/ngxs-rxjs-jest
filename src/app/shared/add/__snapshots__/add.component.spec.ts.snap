// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComponent should match snapshot 1`] = `
<app-add
  store={[Function Store]}
>
  <mat-icon
    class="add-button"
    ng-reflect-menu="[object Object]"
  >
    add_circle
  </mat-icon><mat-menu>
    <button
      mat-menu-item=""
    >
      <mat-icon>
        note_add
      </mat-icon>
      <span>
        Add issue
      </span>
    </button>
    <button
      mat-menu-item=""
    >
      <mat-icon>
        cached
      </mat-icon>
      <span>
        Add random issue
      </span>
    </button>
  </mat-menu>
</app-add>
`;
